tortoisebeacon/votes_calc.go

```diff
@@ -0,0 +1,256 @@
+		ValidVotes:            valid,
+		PotentiallyValidVotes: potentiallyValid,
+	}
+
+	// TODO: also send a bit vector
+	// TODO: initialize margin vector to initial votes
+	// TODO: use weight
+	tb.firstRoundOutcomingVotes[epoch] = votes
+
+	return votes
+}
+
+func (tb *TortoiseBeacon) calcVotes(epoch types.EpochID, round types.RoundID) (votesSetPair, error) {
+	tb.votesMu.Lock()
+	defer tb.votesMu.Unlock()
+
+	// TODO: initialize votes margin when we create a proposal list
+	votesMargin, err := tb.firstRoundVotes(epoch)
+	if err != nil {
+		return votesSetPair{}, fmt.Errorf("calc first round votes: %w", err)
+	}
+
+	tb.Log.With().Debug("Calculated first round votes",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(round)),
+		log.String("votesMargin", fmt.Sprint(votesMargin)))
+
+	ownFirstRoundVotes, err := tb.calcOwnFirstRoundVotes(epoch, votesMargin)
+	if err != nil {
+		return votesSetPair{}, fmt.Errorf("calc own first round votes: %w", err)
+	}
+
+	tb.Log.With().Debug("Calculated own first round votes",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(round)),
+		log.String("votesMargin", fmt.Sprint(votesMargin)),
+		log.String("ownFirstRoundVotes", fmt.Sprint(ownFirstRoundVotes)))
+
+	if err := tb.calcVotesMargin(epoch, round, votesMargin); err != nil {
+		return votesSetPair{}, fmt.Errorf("calc votes count: %w", err)
+	}
+
+	tb.Log.With().Debug("Calculated votes count",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(round)),
+		log.String("votesMargin", fmt.Sprint(votesMargin)))
+
+	ownCurrentRoundVotes, err := tb.calcOwnCurrentRoundVotes(epoch, round, votesMargin)
+	if err != nil {
+		return votesSetPair{}, fmt.Errorf("calc own current round votes: %w", err)
+	}
+
+	tb.Log.With().Debug("Calculated votes for one round",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(round)),
+		log.String("for", fmt.Sprint(ownCurrentRoundVotes.ValidVotes)),
+		log.String("against", fmt.Sprint(ownCurrentRoundVotes.InvalidVotes)))
+
+	return ownCurrentRoundVotes, nil
+}
+
+func (tb *TortoiseBeacon) firstRoundVotes(epoch types.EpochID) (votesMarginMap, error) {
+	firstRoundInThisEpoch := epochRoundPair{
+		EpochID: epoch,
+		Round:   1,
+	}
+
+	// protected by tb.votesMu
+	firstRoundIncomingVotes := tb.incomingVotes[firstRoundInThisEpoch]
+
+	tb.Log.With().Debug("First round incoming votes",
+		log.Uint64("epoch_id", uint64(epoch)),
+		log.Uint64("round", uint64(1)),
+		log.String("votes", fmt.Sprint(firstRoundIncomingVotes)))
+
+	firstRoundVotesMargin := make(map[proposal]int)
+
+	for nodeID, votesList := range firstRoundIncomingVotes {
+		firstRoundVotesFor := make(hashSet)
+		firstRoundVotesAgainst := make(hashSet)
+
+		voteWeight, err := tb.voteWeight(nodeID, epoch)
+		if err != nil {
+			return nil, fmt.Errorf("get vote weight: %w", err)
+		}
+
+		for vote := range votesList.ValidVotes {
+			firstRoundVotesMargin[vote] += int(voteWeight)
```

Comment:
can this overflow?

Reply:
@antonlerner I think so. Should I use `big.Int`?